File: crates/geny-harness-core/src/session/session.rs
//! Single session execution unit.

use chrono::{DateTime, Utc};

use crate::core::config::PipelineConfig;
use crate::core::pipeline::Pipeline;
use crate::core::result::PipelineResult;
use crate::core::state::PipelineState;
use crate::session::freshness::{FreshnessPolicy, FreshnessStatus};

/// Wraps Pipeline + State for multi-turn interactions.
pub struct Session {
    session_id: String,
    pipeline: Pipeline,
    state: PipelineState,
    freshness_policy: FreshnessPolicy,
    created_at: DateTime<Utc>,
    last_active: DateTime<Utc>,
}

impl Session {
    pub fn new(
        session_id: Option<String>,
        pipeline: Pipeline,
        _config: Option<PipelineConfig>,
        freshness_policy: Option<FreshnessPolicy>,
    ) -> Self {
        let now = Utc::now();
        let session_id =
            session_id.unwrap_or_else(|| uuid::Uuid::new_v4().to_string()[..12].to_string());
        let mut state = PipelineState::new();
        state.session_id = session_id.clone();

        Self {
            session_id,
            pipeline,
            state,
            freshness_policy: freshness_policy.unwrap_or_default(),
            created_at: now,
            last_active: now,
        }
    }

    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    pub fn pipeline(&self) -> &Pipeline {
        &self.pipeline
    }

    pub fn state(&self) -> &PipelineState {
        &self.state
    }

    pub fn state_mut(&mut self) -> &mut PipelineState {
        &mut self.state
    }

    /// Evaluate current freshness.
    pub fn freshness(&self) -> FreshnessStatus {
        self.freshness_policy
            .evaluate(self.created_at, self.last_active, self.state.messages.len())
    }

    /// Execute input through pipeline, preserving state.
    pub async fn run(&mut self, input: serde_json::Value) -> PipelineResult {
        self.last_active = Utc::now();
        let state = std::mem::take(&mut self.state);
        let result = self.pipeline.run(input, Some(state)).await;
        // State is consumed by run, create new for next turn
        // In a full implementation, state would be returned/preserved
        self.state = PipelineState::new();
        self.state.session_id = self.session_id.clone();
        result
    }

    /// Clear state for fresh start.
    pub fn reset_state(&mut self) {
        self.state = PipelineState::new();
        self.state.session_id = self.session_id.clone();
    }
}

File: python/geny_harness/session/session.py
"""Session — wraps Pipeline + State for multi-turn interactions."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any, AsyncIterator, Optional

from geny_harness._native import PipelineEvent, PipelineResult, PipelineState
from geny_harness.session.freshness import FreshnessPolicy, FreshnessStatus


class Session:
    """Wraps Pipeline + State for multi-turn interactions."""

    def __init__(
        self,
        session_id: Optional[str] = None,
        pipeline: Any = None,
        config: Any = None,
        freshness_policy: Optional[FreshnessPolicy] = None,
    ):
        self._id = session_id or uuid.uuid4().hex[:12]
        self._pipeline = pipeline
        self._state = PipelineState()
        self._state.session_id = self._id
        self._freshness_policy = freshness_policy or FreshnessPolicy()
        self._created_at = datetime.now(timezone.utc)
        self._last_active = datetime.now(timezone.utc)

        if config is not None:
            config.apply_to_state(self._state)

    @property
    def id(self) -> str:
        return self._id

    @property
    def session_id(self) -> str:
        return self._id

    @property
    def pipeline(self):
        return self._pipeline

    @property
    def state(self) -> PipelineState:
        return self._state

    @property
    def freshness(self) -> FreshnessStatus:
        return self._freshness_policy.evaluate(
            self._created_at, self._last_active, len(self._state.messages)
        )

    async def run(self, input_text: Any) -> PipelineResult:
        """Execute input through pipeline, preserving state."""
        self._last_active = datetime.now(timezone.utc)
        result = await self._pipeline.run(input_text, self._state)
        return result

    async def run_stream(self, input_text: Any) -> AsyncIterator[PipelineEvent]:
        """Streaming execution — yields PipelineEvents."""
        self._last_active = datetime.now(timezone.utc)
        stream = await self._pipeline.run_stream(input_text, self._state)
        async for event in stream:
            yield event

    def reset_state(self) -> None:
        """Clear state for fresh start."""
        self._state = PipelineState()
        self._state.session_id = self._id
